common/src/main/java/org/conscrypt/HpkeAlgorithmIdentifier.java

```diff
@@ -0,0 +1,261 @@
+
+import static org.conscrypt.Preconditions.checkNotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.crypto.Cipher;
+
+/**
+ * KEM, KDF, and AEAD algorithm identifiers and specs defined on RFC 9180.
+ * <ul>
+ *   <li><a href="https://www.rfc-editor.org/rfc/rfc9180.html#name-key-encapsulation-mechanism">KEM</a></li>
+ *   <li><a href="https://www.rfc-editor.org/rfc/rfc9180.html#name-key-derivation-functions-kd">KDF</a></li>
+ *   <li><a href="https://www.rfc-editor.org/rfc/rfc9180.html#name-authenticated-encryption-wi">AEAD</a></li>
+ * </ul>
+ */
+public class HpkeAlgorithmIdentifier {
+
+    public enum KEM {
+        DHKEM_P_256_HKDF_SHA256, DHKEM_P_384_HKDF_SHA384, DHKEM_P_521_HKDF_SHA512, DHKEM_X25519_HKDF_SHA256, DHKEM_X448_HKDF_SHA512;
+
+        private static final Map<KEM, Spec> SPEC_LENGTH_IN_BYTES = new HashMap<>();
+        static {
+            SPEC_LENGTH_IN_BYTES.put(
+                DHKEM_P_256_HKDF_SHA256,
+                new Spec(/* secret= */ 32,/* enc= */ 65,/* pk= */ 65,/* sk= */ 32));
+            SPEC_LENGTH_IN_BYTES.put(
+                DHKEM_P_384_HKDF_SHA384,
+                new Spec(/* secret= */ 48,/* enc= */ 97,/* pk= */ 97,/* sk= */ 48));
+            SPEC_LENGTH_IN_BYTES.put(
+                DHKEM_P_521_HKDF_SHA512,
+                new Spec(/* secret= */ 64,/* enc= */ 133,/* pk= */ 133,/* sk= */ 66));
+            SPEC_LENGTH_IN_BYTES.put(
+                DHKEM_X25519_HKDF_SHA256,
+                new Spec(/* secret= */ 32,/* enc= */ 32,/* pk= */ 32,/* sk= */ 32));
+            SPEC_LENGTH_IN_BYTES.put(
+                DHKEM_X448_HKDF_SHA512,
+                new Spec(/* secret= */ 64,/* enc= */ 56,/* pk= */ 56,/* sk= */ 56));
+        }
+
+        /**
+         * The length in bytes of a KEM shared secret produced by this KEM.
+         *
+         * @return shared secret size in bytes
+         */
+        public int getSecretLength() {
+            return SPEC_LENGTH_IN_BYTES.get(this).secret;
+        }
+
+        /**
+         * The length in bytes of an encapsulated key produced by this KEM.
+         *
+         * @return encapsulated key size in bytes
+         */
+        public int getEncLength() {
+            return SPEC_LENGTH_IN_BYTES.get(this).enc;
+        }
+
+        /**
+         * The length in bytes of an encoded public key for this KEM.
+         *
+         * @return public key size in bytes
+         */
+        public int getPkLength() {
+            return SPEC_LENGTH_IN_BYTES.get(this).pk;
+        }
+
+        /**
+         * The length in bytes of an encoded private key for this KEM.
+         *
+         * @return private key size in bytes
+         */
+        public int getSkLength() {
+            return SPEC_LENGTH_IN_BYTES.get(this).sk;
+        }
+
+        /**
+         * Splits the value that {@link Cipher#doFinal()} returns. The value returned is an
+         * encapsulated key and ciphertext concatenated. This helps splits them apart.
```

Comment:
I'm assuming you all need to concatenate them because you're having to jam it into Java's generic "cipher" API. For BoringSSL, we intentionally opted to build a bespoke HPKE API. I think that's a better model, but I'll defer to you all on Java API design. Wouldn't be the first time Java had a bad crypto API. :-)

Edit: Okay, got a little further along in the API chosen here. This is *definitely* not the right API here. I think this needs another round of API design before it's ready to land. HPKE just doesn't fit in Java's generic "cipher" and needs a different interface, in the same way that a TLS connection is not a "cipher" and has its own interface.